services/powchain/src/main/java/tech/pegasys/teku/services/executionengine/ExecutionEngineService.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.services.executionengine;
+
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.events.EventChannels;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.BeaconBlockBody;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.merge.BeaconBlockBodyMerge;
+import tech.pegasys.teku.spec.datastructures.execution.ExecutionPayload;
+import tech.pegasys.teku.spec.executionengine.ExecutionEngineChannel;
+import tech.pegasys.teku.spec.executionengine.ForkchoicheState;
+import tech.pegasys.teku.storage.api.ChainHeadChannel;
+import tech.pegasys.teku.storage.api.ReorgContext;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class ExecutionEngineService implements ChainHeadChannel {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Spec spec;
+  private final ExecutionEngineChannel executionEngineChannel;
+  private final RecentChainData recentChainData;
+
+  public ExecutionEngineService(
```

Comment:
hmm, I think we're going to have to rethink how we implement `ExecutionEngineService`.  The `powchain` module has so far deliberately avoided making direct access to `recentChainData` as that's held within the beacon chain service and the PowChainService stuff is meant to be separate from that.

And the number of `recentChainData.retrieve` calls in here are potentially quite expensive, and introduce potential for race conditions that we'll need to handle properly.  Most likely using `ChainHeadChannel` updates isn't the right approach as we would have more info available back in `RecentChainData.updateChainHead` that calls this.

Also worth noting that `ChainHeadChannel` doesn't fire any events while we're in long sync mode which is probably not what we want for this case.

It's probably worth pulling this out of this PR for now and doing it in a follow up - I think we've wound up with too much in one PR here.